src/types.rs

```diff
@@ -126,6 +117,29 @@ pub struct CommitmentWitness {
     blinding: Scalar,
 }
 
+/// Representation of a Contract inside an Input that can be cloned.
+#[derive(Clone, Debug)]
+pub struct FrozenContract {
```

Comment:
Question: will we still pass in a `Contract` type to `encode_output`? Or will the verifier have only FrozenContracts and the prover Contracts

Reply:
FrozenContract will be visible only to the prover, and we'll need to implement conversion from FrozenContract into a Contract. Then, the Contract can be encoded per existing rules.